include/alpaka/acc/AccOacc.hpp

```diff
@@ -0,0 +1,514 @@
+/* Copyright 2020 Jeffrey Kelling
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#ifdef ALPAKA_ACC_ANY_BT_OACC_ENABLED
+
+#if _OPENACC < 201306
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
+#endif
+
+// Base classes.
+#include <alpaka/idx/gb/IdxGbOacc.hpp>
+#include <alpaka/idx/bt/IdxBtOacc.hpp>
+#include <alpaka/atomic/AtomicOaccBuiltIn.hpp>
+#include <alpaka/atomic/AtomicHierarchy.hpp>
+#include <alpaka/intrinsic/IntrinsicFallback.hpp>
+#include <alpaka/math/MathStdLib.hpp>
+#include <alpaka/rand/RandStdLib.hpp>
+#include <alpaka/time/TimeStdLib.hpp>
+#include <alpaka/ctx/block/CtxBlockOacc.hpp>
+#include <alpaka/warp/WarpSingleThread.hpp>
+
+// Specialized traits.
+#include <alpaka/acc/Traits.hpp>
+#include <alpaka/kernel/Traits.hpp>
+#include <alpaka/dev/Traits.hpp>
+#include <alpaka/pltf/Traits.hpp>
+#include <alpaka/idx/Traits.hpp>
+
+// Implementation details.
+#include <alpaka/core/ClipCast.hpp>
+#include <alpaka/core/Concepts.hpp>
+#include <alpaka/core/Unused.hpp>
+#include <alpaka/dev/DevOacc.hpp>
+
+#include <alpaka/ctx/block/CtxBlockOacc.hpp>
+
+#include <limits>
+#include <typeinfo>
+
+namespace alpaka
+{
+    template<
+        typename TDim,
+        typename TIdx,
+        typename TKernelFnObj,
+        typename... TArgs>
+    class TaskKernelOacc;
+
+    // define max gang/worker num because there is no standart way in OpenACC to
+    // get this information
+#ifndef ALPAKA_OACC_MAX_GANG_NUM
+    constexpr size_t oaccMaxGangNum = std::numeric_limits<unsigned int>::max();
+#else
+    constexpr size_t oaccMaxGangNum = ALPAKA_OACC_MAX_GANG_NUM;
+#endif
+#if defined(ALPAKA_OFFLOAD_MAX_BLOCK_SIZE) && ALPAKA_OFFLOAD_MAX_BLOCK_SIZE>0
+    constexpr size_t oaccMaxWorkerNum = ALPAKA_OFFLOAD_MAX_BLOCK_SIZE;
+#else
+    constexpr size_t oaccMaxWorkerNum = 1;
+#endif
+
+    //#############################################################################
+    //! The OpenACC accelerator.
+    template<
+        typename TDim,
+        typename TIdx>
+    class AccOacc final :
+        public gb::IdxGbOacc<TDim, TIdx>, // dummy
```

Comment:
We should not inherit from `IdxGbOacc` if it is not doing something. You should delete `IdxGbOacc` completely.

Reply:
The Idea here was, when `AccOacc` and `CtxThreadOacc` were still separate, to make `AccOacc` formally a valid accelerator which implements all concepts, although the once which can only be used in device code are only actually implemented by `CtxThreadOacc`.

`AccOacc` have been a type only used on the host to reference the backend, but which allows concepts implemented by the backend to be checked by host code. `TaskKernelOacc` in any case creates the on-device Acc instance of type `CtxThreadOacc`. Back then I thougth the Acc might be instantiated by some host code, so I kept it in this double way of having `BlockShared` inside classes which cannot be used on the host. 

This can indeed be removed with the `CtyThreadOacc -> AccOacc` merge in place.